src/Workspaces/Core/Portable/CodeRefactorings/FixAllOccurences/DocumentBasedFixAllProvider.cs

```diff
@@ -0,0 +1,216 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+using static Microsoft.CodeAnalysis.FixAll.CommonDocumentBasedFixAllProviderHelpers;
+
+namespace Microsoft.CodeAnalysis.CodeRefactorings
+{
+    using FixAllContexts = Func<FixAllContext, ImmutableArray<FixAllContext>, Task<Solution?>>;
+
+    /// <summary>
+    /// Provides a base class to write a <see cref="FixAllProvider"/> for refactorings that fixes documents independently.
+    /// This type should be used in the case where the code refactoring(s) only affect individual <see cref="Document"/>s.
+    /// </summary>
+    /// <remarks>
+    /// This type provides suitable logic for fixing large solutions in an efficient manner.  Projects are serially
+    /// processed, with all the documents in the project being processed in parallel. 
+    /// <see cref="FixAllAsync(FixAllContext, Document, ImmutableArray{TextSpan})"/> is invoked for each document for implementors to process.
+    ///
+    /// TODO: Make public, tracked with https://github.com/dotnet/roslyn/issues/60703
+    /// </remarks>
+    internal abstract class DocumentBasedFixAllProvider : FixAllProvider
```

Comment:
so, there's a pattern we can apply to prevent the massive amount of duplication happening in this PR.  I get taht you couldn't share code using inheritance because of hte need to subclass a different FixAllProvider in each namespace.  But here's a pattern that can work:

```c#
namespace CodeFixes
{
    class FixAllProvider
    {
    }
  
    class DocumentBasedFixAllProvider : FixAllProvider, ICommonFixAllProvider
    {
        X Something(Y arg1, Z arg2)
            => CommonFixAllProviderHelper.Something(this, arg1, arg2);
    }
}
namespace CodeRefactorings
{
    class FixAllProvider
    {
    }
  
    class DocumentBasedFixAllProvider : FixAllProvider, ICommonFixAllProvider
    {
    }
}

namespaces CodeFixesAndRefactorings
{
    interface ICommonFixAllProvider
    {
    }

    static class CommonFixAllProviderHelper
    {
        static X Something(ICommonFixAllProvider, Y arg1, Z arg2)
        {
            // do common logic here.
        }
    }
}
```

You can also use generics here on CommonFixAllProviderHelper (and CommonFixAllProviderHelper.Something) with constraints to get strong type safety if you need it (for example, for things like the different FixAllContexts and whatnot).

The critical thing here is that CommonFixAllProviderHelper is where all the code exists (and can be updated just once), and the two DocumentBasedFixAllProvider just defer to it.  The implement ICommonFixAllProvider so that the helper type can do whatever it needs with teh original type without caring which subclass it is.

Reply:
Going to give it a try to share more code here. The tricky part was that existing FixAll related types are public and cannot have an internal base type, but probably just using common interfaces and shared static helper class for common code would work.